core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -36,373 +73,288 @@ class Queue[A] private (
-            )
-            .flatMap(_.join)
-          (forked *> hook, Shutdown)
-        } else (hook, Shutdown)
-      case Shutdown => (IO.unit, Shutdown)
-    })
+  final val shutdown: IO[Nothing, Unit] = for {
+    hook <- shutdownHook.modify {
+             case None       => (IO.unit, None)
+             case Some(hook) => (hook, None)
+           }
+    takers <- IO.sync(unsafePollAll(takers))
+    _      <- IO.parTraverse(takers)(_.interrupt) *> hook
+    _      <- strategy.shutdown
+  } yield ()
 
   /**
    * Removes the oldest value in the queue. If the queue is empty, this will
    * return a computation that resumes when an item has been added to the queue.
    */
-  final val take: IO[Nothing, A] = {
-
-    val acquire: (Promise[Nothing, A], State[A]) => (IO[Nothing, Boolean], State[A]) = {
-      case (p, Deficit(takers, hook)) => (IO.now(false), Deficit(takers.enqueue(p), hook))
-      case (p, Surplus(values, putters, hook)) =>
-        strategy match {
-          case Sliding | Dropping if capacity.exists(_ < 1) =>
-            (IO.never, Surplus(IQueue.empty, putters, hook))
-          case _ =>
-            values.dequeueOption match {
-              case None if putters.isEmpty =>
-                (IO.now(false), Deficit(IQueue.empty.enqueue(p), hook))
-              case None =>
-                val (newSurplus, promise) =
-                  moveNPutters(Surplus(values, putters, hook), 1)
-                newSurplus.queue.dequeueOption match {
-                  case None => (promise *> IO.now(false), newSurplus)
-                  case Some((a, values)) =>
-                    (
-                      promise *> p.complete(a),
-                      Surplus(values, newSurplus.putters, hook)
-                    )
-                }
-              case Some((a, values)) =>
-                val (newSurplus, promise) =
-                  moveNPutters(Surplus(values, putters, hook), 1)
-                (promise *> p.complete(a), newSurplus)
-
-            }
-        }
-      case (p, state @ Shutdown) => (p.interrupt, state)
-    }
-
-    val release: (Boolean, Promise[Nothing, A]) => IO[Nothing, Unit] = {
-      case (_, p) => p.poll.void <> removeTaker(p)
-    }
-
-    Promise.bracket[Nothing, State[A], A, Boolean](ref)(acquire)(release)
-
-  }
+  final val take: IO[Nothing, A] =
+    for {
+      _    <- checkShutdownState
+      item <- IO.sync(queue.poll(null.asInstanceOf[A]))
+      res <- if (item != null) strategy.onQueueEmptySpace(queue).map(_ => item)
+            else
+              for {
+                p <- Promise.make[Nothing, A]
+                // add the promise to takers, then try take again in case a value was added since
+                _ <- IO.sync(takers.offer(p)) *> completeTakers
+                // wait for the promise to be completed, and clean up resources in case of interruption
+                res <- p.get.ensuring(p.poll.void <> removeTaker(p))
```

Comment:
Unfortunately, there is a memory leak here:

The finalizer is added to `p.get`. So if the take action is interrupted before `p.get`, then the taker will never be removed.

However, you can solve this quite easily:

```scala
for {
  p <- Promise.make[Nothing, A]
  a <- (IO.sync(takers.offer(p)) *> completeTakers).flatMap(p.get).ensuring(p.poll.void <> removeTaker(p))
} yield a
```

This pattern (or a variant) should be used at all times to ensure safe cleanup of takers / putters.

Reply:
Indeed! I changed for both takers and putters.